clients/pkg/promtail/wal/watcher.go

```diff
@@ -0,0 +1,287 @@
+package wal
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wlog"
+)
+
+const (
+	readPeriod         = 10 * time.Millisecond
+	segmentCheckPeriod = 100 * time.Millisecond
+)
+
+// Based in the implementation of prometheus WAL watcher
+// https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/watcher.go. Includes some changes to make it suitable
+// for log WAL entries, but also, the writeTo surface has been implemented according to the actions necessary for
+// Promtail's WAL.
+
+// Reader is a dependency interface to inject generic WAL readers into the Watcher.
+type Reader interface {
+	Next() bool
+	Err() error
+	Record() []byte
+}
+
+// WriteTo is responsible for doing the necessary work to process both series and entries while the Watcher
+// is reading / tailing segments. The implementor of this interface is not expected to implement thread safety if used
+// on a single watcher, since each callback is called synchronously.
+//
+// Based on https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/watcher.go#L46
+type WriteTo interface {
+	StoreSeries(series record.RefSeries)
+	Append(entries RefEntries) error
+}
```

Comment:
this interface will likely still need the cleanup functions from prometheus' `WriteTo` interface

Reply:
Adding for cleanup just `SeriesReset(int)` for now. Since alongside with the segment deletion mechanism the WAL writer uses, we can use something similar that the watcher somehow notifies that a segment that was seen before is now gone, so the WriteTo can clean up unused series